lib/srv/db/postgres/test.go

```diff
@@ -297,7 +365,53 @@ func (s *TestServer) ParametersCh() chan map[string]string {
 
 // Close closes the server listener.
 func (s *TestServer) Close() error {
-	return s.listener.Close()
+	closeErr := s.listener.Close()
+	s.cleanupPids()
+	return trace.Wrap(closeErr)
+}
+
+// newPid makes a unique PID and inserts it into the server's pid map.
+// For test purposes, every pid will have the same stub secret key.
+func (s *TestServer) newPid() uint32 {
+	s.pidMu.Lock()
+	defer s.pidMu.Unlock()
+	s.nextPid++
+	s.pids[s.nextPid] = &pidHandle{secretKey: testSecretKey}
+	return s.nextPid
+}
+
+// cleanupPid cleans up a pid by calling its cancel func and
+// deleting it from the pid map.
+func (s *TestServer) cleanupPid(pid uint32) {
+	s.pidMu.Lock()
+	defer s.pidMu.Unlock()
+	if entry, ok := s.pids[pid]; ok && entry != nil && entry.cancel != nil {
+		entry.cancel()
+		delete(s.pids, pid)
```

Comment:
nit: should we delete even if cancel is nil?

Reply:
I think we might as well, although it doesn't really matter since PIDs are monotonically increasing to keep it simpler. I mainly wanted to prevent context leaks from failing to call cancel